### ISSUE TITLE
GraphQL POST requests are not cached

### ISSUE THREAD

**Issue by viters** [directus/directus#11418]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

GraphQL requests will not be cached nor returned from cache when using `directus/sdk` or Directus docs information. 

Directus implementation caches only GET requests:
`api/src/middleware/respond.ts:17`
![image](https://user-images.githubusercontent.com/13100466/152409073-e59b8fa7-1e88-4f75-8a3a-64014be7c496.png)

Also, cache key for GraphQL post requests will be always the same irregardless of variables / query:
`api/src/utils/get-cache-key.ts:13`
![image](https://user-images.githubusercontent.com/13100466/152409171-ae25d860-9909-42ce-951b-256c859b709f.png)

Contrary to docs (where `POST /graphql` is suggested), implementation of `api/src/middleware/graphql.ts` requires (or suggests) GET requests to fetch data. 

### To Reproduce

1. Run Directus
2. Use SDK or use `POST /graphql` to get data
3. Cache is not used at all (for example monitor redis).

### What version of Directus are you using?

9.4.3 (but code does not change for 9.5.0)

### What version of Node.js are you using?

N/A

### What database are you using?

N/A

### What browser are you using?

N/A

### What operating system are you using?

N/A

### How are you deploying Directus?

N/A

**Comment by rijkvanzanten** [comment]
What do you mean by

> Also, cache key for GraphQL post requests will be always the same irregardless of variables / query:

What I see in the highlighted line is that it's using GraphQL's query+variables when it's graphql request, and the REST sanitizedQuery when not.

**Comment by viters** [comment]
@rijkvanzanten 
`req.query` is {} for `POST /graphql`. `req.query.query` and `req.query.variables` is undefined.

**Comment by viters** [comment]
The workaround for now is to use GET requests, probably Directus SDK can be somehow overwritten to use GET instead of POST (like by passing transport that on `.post` call sends GET request).

**Comment by rijkvanzanten** [comment]
True 👍🏻 

Also keep in mind that—regardless of Directus' internal caches—browsers won't cache anything on POSTs either. This is a known issue with GraphQL as a tool: https://www.apollographql.com/blog/backend/caching/graphql-caching-the-elephant-in-the-room/

**Comment by viters** [comment]
@rijkvanzanten 
Yeah, I do not mind GET's actually - although it probably should be configurable to use either GET or POST on SDK, and Directus API should support both (with Directus cache as well). The difference between using GET / POST could be then mentioned in docs (although I do not know about any problems with `GET /graphql` queries and they work fine in Directus, even with variables).

The problem right now is that SDK and Docs recommend POST, while it is a huge performance downgrade in comparison with GET.

**Comment by rijkvanzanten** [comment]
Yeah totally. The internal cache should obviously work for both 👍🏻

**Comment by viters** [comment]
Right now I have a summary:

To support cache for `POST /graphql` queries we would need:
- another condition route in `api/src/middleware/respond.ts:16`, where we check for `POST`, `/graphql` url and we must also verify that it is query, not mutation
- `api/src/utils/get-cache-key.ts:13` should also have a route for that use case

I dunno yet about `api/src/middleware/graphql.ts`, there is a bunch of code only for `GET` requests, but `POST /graphql` queries work just fine, so it's hard to tell for me right now if there is anything to be changed in that file.

Optionally, SDK could have some configuration to use GET for queries instead of POST.

**Comment by rijkvanzanten** [comment]
> GET could have a limitation of max query length (url length).

Yeah. There's an in-works spec to make an official ~~SEARCH~~ QUERY HTTP method that works around that, by effectively being a GET request with a body (ref https://github.com/directus/directus/discussions/11362)

**Comment by viters** [comment]
@rijkvanzanten 

I will work on it over the weekend.

**Comment by viters** [comment]
@rijkvanzanten 
I have opened PR. I have reviewed `parseGraphQL` in `api/src/middleware/graphql.ts` and no changes are required there. The only doubt I had was in `api/src/middleware/cache.ts` and I've put it into a comment, although I think the solution I proposed is fine.

BTW. Small suggestion: add `.git/` to testing Dockerfile's `.dockerignore` (in main dir), a lot less context will be passed to build.
